src/lib/device/virtual_dev.cpp

```diff
@@ -434,7 +438,8 @@ BlkAllocStatus VirtualDev::do_alloc_blk(blk_count_t nblks, const blk_alloc_hints
         BlkAllocStatus status = BlkAllocStatus::FAILED;
 
         while (try_streams != 0) {
-            preferred_chunk = reinterpret_cast< PhysicalDevChunk* >(stream_info->chunk_list[stream_info->stream_cur]);
+            if(m_chunk_selector) preferred_chunk = m_chunk_selector->select_chunk();
```

Comment:
maybe select_chunk need take some argument at least hints.

Reply:
actually I do not think this is the usage pattern for HeapChunkSelector.  this function will be called for per alloc blkc request and if select_chunk() is called here, the chunk will be remove from the HeapChunkSelector and these is no chance for the blk alloc request from the shard to allocate blk from this chunk anymore. 

Instead I think select_chunk() will be called from honeobject layer and once chunk is slected, it will be used as a hint to alloc blk to tell it use this chunk as destination to alloc blk

Reply:
> the chunk will be remove from the HeapChunkSelector and these is no chance for the blk alloc request from the shard to allocate blk from this chunk anymore.

when we close a shard, the chunk is not busy, so we should call add_chunk() to return it back to chunkselector . so that this chunk can be reused for the next blk allocate

Reply:
yes, that is sure. but my point is before shard is sealed/closed,   put shard request will try to allocate a blk from honestore, but now the first put request of this shard, a chunk will be selected and chunk is removed from selector.      let us say that is chunkA. after that,  the next put request of this shard will try to allocate a new block with this function, but it will return another chunk again which is definitely not chunk A.

Reply:
if i don`t missunderstand, put shard request is actually the create shard request, which will put a block as a shard header, right?

if yes , when a shard header is put to the chunk, this means the shard is open and the chunk is in busy state, we could not return the chunk to the selector until the shard is sealed. so next put request will also try to allocate a new block in the same chunk.

Reply:
no,  put shard mean put blobs to the shard, both put shard and create shard will allocate blk from data service, but create shard will put shard header to the blk, but put shard will put blob content to the blk.  both of them will need to allocate blk.

Reply:
yea, get your point here. you mean we should expose `select_chunk()` to homeobject, not in `allocate_block`, so that homeobject can manage the chunk itself. it will return the chunk to chunkselector until homeobject seal a shard. homeobject will manager blocks itself when allocating a block, not call the `allocateBlk` of homestore , right?

Reply:
> yea, get your point here. you mean we should expose `select_chunk()` to homeobject, not in `allocate_block`, so that homeobject can manage the chunk itself. it will return the chunk to chunkselector until homeobject seal a shard. 

yes, that is my thought. 

> homeobject will manager blocks itself when allocating a block, not call the `allocateBlk` of homestore , right?
that is not decided yet. I think both work.  it can allocate a block from the selected chunk directly or call allocateBlk but with chunk as the hint params.   

let us hear more thoughts from other workmates.

Reply:
I hope we can still leave the `alloc_blks` in homestore and homeobj still use that.

@zichanglai the shard->chunk affinitiy should be implemented in `HeapChunkSelector::select_chunk(const blk_alloc_hints& hints)` in `HeapChunkSelector'.

The most obvious way is add the `chunk` into alloc_hint , in this case `HeapChunkSelector::select_chunk` always return that chunk in the alloc_hint`.  And `alloc_blks` return chunk info to caller(homeobj).  But this approach is ugly.

Would prefer we add `shardID` , `PGID` into  `alloc_hint`,  `HeapChunkSelector::select_chunk` can maintain a mapping table  to ensure same `shardID` always get same chunk and same `PGID` always return chunks within same device.

Reply:
>  The most obvious way is add the chunk into alloc_hint , in this case HeapChunkSelector::select_chunk always return that chunk in the alloc_hint.  And alloc_blks` return chunk info to caller(homeobj). But this approach is ugly

if we add chunk to the alloc_hint,  there is no need to call HeapChunkSelector any interface any more. it will just call chunk->blk_allocator_mutable()->alloc() interface to allocate a new blk.

> Would prefer we add shardID , PGID into alloc_hint, HeapChunkSelector::select_chunk can maintain a mapping table to ensure same shardID always get same chunk and same PGID always return chunks within same device.

I am ok with this, but something lilely "shardID" and "PGID" are something homestore doest not know at all, but blk_alloc_hints is a homestore struct, it will be better to called user data area or something else. and for homeobject, the user data area will store shard_id/pg_id.

Reply:
yes that makes sense.

Can we move this discussion to slack to hear more from others?